mpc/mpc-aio/src/protocol/garble/exec/deap/leader.rs

```diff
@@ -0,0 +1,310 @@
+                ev_labels,
+                input_state: InputState {
+                    ot_receive_inputs: ot_receive_inputs
+                        .into_iter()
+                        .map(|v| v.group().clone())
+                        .collect(),
+                },
+            },
+            circ: self.circ,
+            channel: self.channel,
+            backend: self.backend,
+            label_sender: self.label_sender,
+            label_receiver: self.label_receiver,
+        })
+    }
+}
+
+impl<B, LS, LR> DEAPLeader<LabelSetup, B, LS, LR>
+where
+    B: Generator + Evaluator + Compressor + Validator + Send,
+    LS: ObliviousSend<FullInputLabels> + Send,
+    LR: ObliviousReceive<InputValue, ActiveInputLabels> + ObliviousVerify<FullInputLabels> + Send,
+{
+    /// Execute first phase of the protocol, returning the authenticated output.
+    pub async fn execute(
+        self,
+    ) -> Result<(Vec<OutputValue>, DEAPLeader<Executed, B, LS, LR>), GCError> {
+        // Discard summary
+        let (output, _, leader) = self.execute_and_summarize().await?;
+        Ok((output, leader))
+    }
+
+    /// Execute first phase of the protocol, returning the authenticated output
+    /// and a summary of the follower's garbled circuit.
+    ///
+    /// This can be used when the labels of the evaluated circuit are needed.
+    pub async fn execute_and_summarize(
+        mut self,
+    ) -> Result<
+        (
+            Vec<OutputValue>,
+            GarbledCircuit<gc_state::EvaluatedSummary>,
+            DEAPLeader<Executed, B, LS, LR>,
+        ),
+        GCError,
+    > {
+        let leader = core::DEAPLeader::new(self.circ.clone());
+
+        // Garble circuit
+        let full_gc = self
+            .backend
+            .generate(self.circ.clone(), self.state.gen_labels)
+            .await?;
+
+        let (partial_gc, leader) = leader.from_full_circuit(full_gc)?;
+
+        // Send garbled circuit to follower
+        self.channel
+            .send(GarbleMessage::GarbledCircuit(partial_gc.into()))
+            .await?;
+
+        // Receive follower's garbled circuit
+        let msg = expect_msg_or_err!(
+            self.channel.next().await,
+            GarbleMessage::GarbledCircuit,
+            GCError::Unexpected
+        )?;
+
+        // Check their gc against circuit spec
+        let gc_ev =
+            GarbledCircuit::<gc_state::Partial>::from_unchecked(self.circ.clone(), msg.into())?;
+
+        // Evaluate follower's garbled circuit
+        let gc_evaluated = self.backend.evaluate(gc_ev, self.state.ev_labels).await?;
+
+        // Summarize follower's circuit to reduce memory footprint
+        let gc_evaluated_summary = gc_evaluated.get_summary();
+
+        // Also compress follower's circuit to reduce memory footprint before validating it in the next phase
+        let gc_cmp = self.backend.compress(gc_evaluated).await?;
+
+        // Commit to output of follower's circuit
```

Comment:
It's better to use DualEx terminilogy here: we are committing to the value from the equality test of DualEx.